### DISCUSSION TITLE
Compose Multiplatform distributable for Windows doesn't run

### DISCUSSION THREAD

**Discussion by csandels** [hydraulic-software/conveyor#100] [Q&A]
I'm unable to get conveyor on Mac to build a distributable package for Windows 11.  I'm not trying to sign it.  I really just want to run the application on Windows, which is configured to allow apps from anywhere to run.  I can use "conveyor make site" and I'll get a zip file that contains a bunch of files, including a bin directory with the exe in it.  I believe I should be able to unzip this file on the windows machine and run the exe shouldn't I?

When I try to run the exe, I get a dialog that pops up with an error saying "java.lang.ExceptionInInitializerError @ android.compose.ui.awt.WindowComposeBridge.<init>(WindowComposeBridge.desktop.kt:43)"

My project is the default starting project for jetpack compose as downloaded from the Kotlin Multiplatform Wizard at https://kmp.jetbrains.com with only Desktop selected.

My build file looks like this:

```
import org.jetbrains.compose.ExperimentalComposeLibrary
import org.jetbrains.compose.desktop.application.dsl.TargetFormat

plugins {
    alias(libs.plugins.kotlinMultiplatform)
    alias(libs.plugins.jetbrainsCompose)
    id("dev.hydraulic.conveyor") version "1.6"
    java
}

group = "com.example"
version = "1.0"

kotlin {
    jvm("desktop") {
        withJava()
    }
    jvmToolchain(17)
    
    sourceSets {
        val desktopMain by getting
        
        commonMain.dependencies {
            implementation(compose.runtime)
            implementation(compose.foundation)
            implementation(compose.material)
            implementation(compose.ui)
            @OptIn(ExperimentalComposeLibrary::class)
            implementation(compose.components.resources)
        }
        desktopMain.dependencies {
            implementation(compose.desktop.currentOs)
        }
    }
}


compose.desktop {
    application {
        mainClass = "MainKt"

//        nativeDistributions {
//            targetFormats(TargetFormat.Dmg, TargetFormat.Msi, TargetFormat.Deb)
//            packageName = "org.example.project"
//            packageVersion = "1.0.0"
//        }
    }
}

//dependencies {
//    // Use the configurations created by the Conveyor plugin to tell Gradle/Conveyor where to find the artifacts for each platform.
////    linuxAmd64(compose.desktop.linux_x64)
////    macAmd64(compose.desktop.macos_x64)
////    macAarch64(compose.desktop.macos_arm64)
////    implementation(compose.desktop.currentOs)
//    windowsAmd64(compose.desktop.windows_x64)
//}
```

My conveyor.conf looks like this:

```
include "#!../gradlew -q printConveyorConfig"

// This enables native library extraction, which improves app startup time and robustness.
// It's optional but a good idea to have it. You can remove this if you get tired of specifying
// system properties for Java libraries with native components.
//
// See https://hydraulic.dev/blog/11-in-jar-signing.html for more.
include required("https://raw.githubusercontent.com/hydraulic-software/conveyor/master/configs/jvm/extract-native-libraries.conf")

// Config file documentation: https://conveyor.hydraulic.dev/latest/configs
app {
  // The base URL is where the download site will be placed. Packages will check here for updates.
  site.base-url = "localhost:3000"

  // This is optional. If not specified the last component of the rdns name will be turned into a display name.
  display-name = "WinTest"

  // A global identifier for the app.
  rdns-name = "org.example.project"

  machines = [windows.amd64]
//  machines = [windows.aarch64]

  // Use the Compose logo as an icon and include it so we can pass it to the Window constructor.
  icons = icons/icon.svg
#   windows.inputs += TASK/rendered-icons/windows
#   linux.inputs += TASK/rendered-icons/linux

  // For iteration speed. Remove for release.
  compression-level = low

  updates = none
}

conveyor.compatibility-level = 12
```

I had a hunch maybe the issue is that the dependencies aren't calling out the windows platform but instead are using currentOs, which is Mac.  But when I use the windowsAmd64(compose.desktop.windows_x64) instead, then skiko lib can't be found, among others, and basically the build falls apart with missing dependencies.

**Comment by leo-hydraulic** [comment]
Hello,

A common problem with Compose desktop projects is due to https://github.com/JetBrains/compose-jb/issues/1404
Conveyor's Compose example (which you get by running `conveyor generate compose $SOME_RDNS`) contains the following snippet at the end of build.gradle.kts:

```
configurations.all {
    attributes {
        // https://github.com/JetBrains/compose-jb/issues/1404#issuecomment-1146894731
        attribute(Attribute.of("ui", String::class.java), "awt")
    }
}
```

That should make it work for you. I managed to successfully build a project with the same files you provide here.

**Comment by csandels** [comment]
Thank you for the response Leo.  That does fix the problem.  Also, another way to fix it I discovered was to change the desktopMain dependencies to use windows_x64 instead of currentOs.  I think I ran into trouble when trying to do that in the root dependencies block rather than desktopMain.
```
desktopMain.dependencies {
    implementation(compose.desktop.windows_x64)
}
```
